internal/saucecloud/retry/basicretrier.go

```diff
@@ -1,17 +1,168 @@
 package retry
 
 import (
+	"context"
 	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
 
 	"github.com/rs/zerolog/log"
+	"github.com/saucelabs/saucectl/internal/cypress"
 	"github.com/saucelabs/saucectl/internal/job"
+	"github.com/saucelabs/saucectl/internal/msg"
+	"github.com/saucelabs/saucectl/internal/playwright"
+	"github.com/saucelabs/saucectl/internal/progress"
+	"github.com/saucelabs/saucectl/internal/saucecloud/zip"
+	"github.com/saucelabs/saucectl/internal/saucereport"
+	"github.com/saucelabs/saucectl/internal/storage"
+	"github.com/saucelabs/saucectl/internal/testcafe"
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
 )
 
-type BasicRetrier struct{}
+type BasicRetrier struct {
+	VDCReader       job.Reader
+	ProjectUploader storage.AppService
 
-func (b *BasicRetrier) Retry(jobOpts chan<- job.StartOptions, opt job.StartOptions, previous job.Job) {
+	CypressProject    cypress.Project
+	PlaywrightProject playwright.Project
+	TestcafeProject   testcafe.Project
+}
+
+func (r *BasicRetrier) Retry(jobOpts chan<- job.StartOptions, opt job.StartOptions, previous job.Job) {
+	if r.VDCReader != nil && opt.SmartRetry.FailedTestsOnly {
+		r.RetryFailedTests(jobOpts, opt, previous)
+		return
+	}
+
+	log.Info().Str("suite", opt.DisplayName).
+		Str("attempt", fmt.Sprintf("%d of %d", opt.Attempt+1, opt.Retries+1)).
+		Msg("Retrying suite.")
+	jobOpts <- opt
+}
+
+func (r *BasicRetrier) RetryFailedTests(jobOpts chan<- job.StartOptions, opt job.StartOptions, previous job.Job) {
+	failedTests, err := r.getFailedTests(previous)
+	if err != nil {
+		log.Debug().Err(err).Msgf(msg.UnableToFetchFile, saucereport.SauceReportFileName)
+		log.Info().Msg(msg.SkippingSmartRetries)
+		jobOpts <- opt
+		return
+	}
+	tempDir, err := os.MkdirTemp(os.TempDir(), "saucectl-app-payload-")
+	if err != nil {
+		log.Debug().Err(err).Msgf(msg.UnableToFetchFile, saucereport.SauceReportFileName)
+		log.Info().Msg(msg.SkippingSmartRetries)
+		jobOpts <- opt
+		return
+	}
+
+	var runnerFile string
+	var project interface{}
+	switch opt.Framework {
+	case cypress.Kind:
+		r.CypressProject.SetTestGrep(opt.SuiteIndex, failedTests)
+		project = r.CypressProject
+	case playwright.Kind:
+		r.PlaywrightProject.Suites[opt.SuiteIndex].Params.Grep = strings.Join(failedTests, "|")
+		project = r.PlaywrightProject
+	case testcafe.Kind:
+		r.TestcafeProject.Suites[opt.SuiteIndex].Filter.TestGrep = strings.Join(failedTests, "|")
+		project = r.TestcafeProject
```

Comment:
Some general thoughts:

1. How much framework specific knowledge do we want to have in the retrier? It obviously has to have some, since it needs to reconfigure the starting options. But how the params are concatenated. Is that something the retrier should know? Or rather just tell the project "here are the tests for this suite" (the way you did for cypress).

2. 🍝 with regards to archiving the runner config. Now we have the `saucecloud` package that had this _private_ knowledge, which you extracted into `zip`, as well as the `retrier` take part of this workflow. Maybe that's the best we can do or maybe there's a better way?